core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.encryption;
+
+import java.nio.ByteBuffer;
+import java.security.SecureRandom;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * Encryption manager that performs envelope encryption.
+ */
+public class EnvelopeEncryptionManager implements EncryptionManager {
+  private final EnvelopeConfig dataEncryptionConfig;
+  private final boolean pushdown;
+  private final KmsClient kmsClient;
+  private final int dataKeyLength;
+
+  private transient volatile SecureRandom workerRNG = null;
+
+  public EnvelopeEncryptionManager(
```

Comment:
nit: style - one line

Reply:
This seems like it has enough that we can have a unit test which encrypts and decrypts a file? Would there be some values in having that? And also triggering with and without a Kms supporting key generation?

Reply:
Probably needs Javadoc

Reply:
> This seems like it has enough that we can have a unit test which encrypts and decrypts a file? Would there be some values in having that? And also triggering with and without a Kms supporting key generation?

Actual file encryption / decryption would require both a Kms and a "native" stack (eg Parquet). These are defined in different PRs. There is a unitest in a later PR that involves these parts and tests the EnvelopeEncryptionManager in a file encryption / decryption.